go/libkb/per_user_key.go

```diff
@@ -0,0 +1,598 @@
+func newPerUserKeyPrev(contents PerUserKeySeed, symmetricKey NaclSecretBoxKey) (PerUserKeyPrev, error) {
+	if contents.IsBlank() {
+		return "", errors.New("attempt to secretbox blank per-user-key")
+	}
+
+	const version = 1
+
+	var nonce [NaclDHNonceSize]byte
+	if nRead, err := rand.Read(nonce[:]); err != nil {
+		return "", err
+	} else if nRead != NaclDHNonceSize {
+		return "", fmt.Errorf("Short random read: %d", nRead)
+	}
+
+	// secretbox
+	var symmetricKey2 [NaclSecretBoxKeySize]byte = symmetricKey
+	sealed := secretbox.Seal(nil, contents[:], &nonce, &symmetricKey2)
+
+	parts := []interface{}{version, nonce, sealed}
+
+	// msgpack
+	mh := codec.MsgpackHandle{WriteExt: true}
+	var msgpacked []byte
+	enc := codec.NewEncoderBytes(&msgpacked, &mh)
+	err := enc.Encode(parts)
+	if err != nil {
+		return "", err
+	}
+
+	// b64
+	return PerUserKeyPrev(base64.StdEncoding.EncodeToString(msgpacked)), nil
+}
+
+// Opens the output of NewPerUserKeyPrev
+func openPerUserKeyPrev(sbox string, symmetricKey NaclSecretBoxKey) (PerUserKeySeed, error) {
+	var res PerUserKeySeed
+
+	// decode b64
+	msgpacked, err := base64.StdEncoding.DecodeString(sbox)
+	if err != nil {
+		return res, err
+	}
+
+	// decode msgpack
+	mh := codec.MsgpackHandle{WriteExt: true}
+	dec := codec.NewDecoderBytes(msgpacked, &mh)
+	var parts struct {
+		Version int
+		Nonce   []byte
+		Sealed  []byte
+	}
+	err = dec.Decode(&parts)
+	if err != nil {
+		return res, err
+	}
+
+	// check parts
+	if parts.Version != 1 {
+		return res, fmt.Errorf("per user key secret box version %v != 1", parts.Version)
+	}
+
+	if len(parts.Nonce) != NaclDHNonceSize {
+		return res, fmt.Errorf("per user key secret box nonce length %v != %v",
+			len(parts.Nonce), NaclDHNonceSize)
+	}
+	var nonce [NaclDHNonceSize]byte
+	copy(nonce[:], parts.Nonce)
+
+	expectedSealedLength := PerUserKeySeedSize + secretbox.Overhead
+	if len(parts.Sealed) != expectedSealedLength {
+		return res, fmt.Errorf("per user key secret box sealed length %v != %v", len(parts.Sealed), expectedSealedLength)
+	}
+
+	// open secretbox
+	var symmetricKey2 [NaclSecretBoxKeySize]byte = symmetricKey
+	contents, ok := secretbox.Open(nil, parts.Sealed, &nonce, &symmetricKey2)
+	if !ok {
+		return res, errors.New("per user key secret box open failed")
+	}
+	if len(contents) != PerUserKeySeedSize {
+		return res, fmt.Errorf("per user key seed length %v != %v", len(contents), PerUserKeySeedSize)
+	}
+
+	res, err = MakeByte32Soft(contents)
+	if err != nil {
+		return res, err
+	}
+	return res, nil
+}
+
+type perUserKeyFull struct {
+	seed   PerUserKeySeed
+	sigKey *NaclSigningKeyPair
+	encKey *NaclDHKeyPair
+}
+
+type PerUserKeyMap map[keybase1.PerUserKeyGeneration]perUserKeyFull
+
+// PerUserKeyring holds on to all versions of the per user key.
+// Generation=0 should be nil, but all others should be present.
+type PerUserKeyring struct {
+	Contextified
+	sync.Mutex
+	uid         keybase1.UID
+	generations PerUserKeyMap
+}
```

Comment:
Not part of this PR, but what are our intentions for disk caching these keys? You mentioned @maxtaco was skeptical about it?